k8srm-prototype/pkg/api/claim_types.go

```diff
@@ -193,129 +194,36 @@ type DeviceClaimStatus struct {
 	// +optional
 	PodNames []string `json:"podNames,omitempty"`
 }
 
-// DeviceSetClaim is one idea for how we can do "composite devices". Rather
-// than changing how the devices are published by drivers, we just change
-// they are consumed. This allows us to combine claims into a group, and apply
-// additional constraints across the group.
-type DeviceSetClaim struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec   DeviceSetClaimSpec   `json:"spec,omitempty"`
-	Status DeviceSetClaimStatus `json:"status,omitempty"`
-}
-
-// DeviceSetClaimSpec contains the details for requesting a set of devices
-// as a unit.
-type DeviceSetClaimSpec struct {
-	// TODO: Maybe we want DeviceSetClass ?
-
-	// MatchAttributes allows specifying a constraint that will apply
-	// across all of the claims. For example, if you specified "numa",
-	// then this claim could only be successfully fulfilled if all of
-	// the included claims could be fulfilled by pools with the same
-	// "numa" attribute value. If we simply set matchAttributes in each
-	// claim separately, then they could be consistent within claims, but
-	// inconsistent across claims. Therefore, we need this additional
-	// resource.
-	//
-	// +optional
-
-	MatchAttributes []string `json:"matchAttributes,omitempty"`
-
-	ClaimSpec []DeviceClaimSpec `json:claimSpec,omitempty"`
-}
-
-type DeviceSetClaimStatus struct {
-	ClaimStatus []DeviceClaimStatus `json:"claimStatus,omitempty"`
-}
-
-// DeviceClaimTemplate is used to generate claims along with Pods. These
-// generated claims have the same lifecycle as the Pod.
-// TODO: Could we just use a DeviceClaim here? Or is that too confusing?
-type DeviceClaimTemplate struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec DeviceClaimTemplateSpec `json:"spec,omitempty"`
-}
-
-// DeviceClaimTemplateSpec contains the information needed to generate
-// DeviceClaims, DevicePrivilegedClaims, or DeviceSetClaims.
-type DeviceClaimTemplateSpec struct {
-	metav1.ObjectMeta
-
-	// Exactly one of these must be populated
-	DeviceClaimSpec           *DeviceClaimSpec           `json:"claimSpec,omitempty"`
-	DevicePrivilegedClaimSpec *DevicePrivilegedClaimSpec `json:"privilegedClaimSpec,omitempty"`
-	DeviceSetClaimSpec        *DeviceSetClaimSpec        `json:"setClaimSpec,omitempty"`
-}
+// NOTE: We no longer have DeviceClaimTemplate. Instead, the PodSpec will
+// directly contain a either a DeviceClaimName (to enable multiple pods to
+// refer to a pre-provisioned claim), or an embedded struct that includes
+// ObjectMeta and a list of the *unrolled fields* of DeviceClaimSpec.  The
+// DeviceClaimSpec type itself will not be embedded, but instead its fields
+// duplicated, allowing them to evolve independently (and have independent
+// validation and avoid Go cyclical dependencies).
+//
+// NOTE: Feedback on this plan has been negative; the complexity of claims may
```

Comment:
for the people without the historical context, the feedback was negative on DeviceClaimTemplate or the alternative proposal?

Reply:
We had no negative feedback on claim templates back when we originally proposed DRA. What we got was a big "no way" when we suggested to embed the template directly in the PodSpec, because it would have been impossible to evolve the content of that template through versioning (there's just a "v1").

I think the same objection still applies to any proposal that depends on putting more than a few trivial fields into the v1 core API.

What is possible is to evolve some new API in a separate API group and once it reaches v1, then embed that structure in a PodSpec.

Reply:
Sorry, I should remove this comment. This was based on very limited feedback *of this API* (not existing 1.30 stuff) from a small group of people who think specifying claims will be too hard for workload authors, and so the complexity of that needs to move to the class. But I have other feedback that is exactly the opposite, that the workload authors know exactly what they want and understand this level of detail. I suspect both are true so we have to accommodate both...